### PR TITLE
Allow `pallet-election-provider` to accept smaller solutions as well

### LINKED ISSUE #9478: Allow pallet-election-provider to accept smaller solutions as well
This is taken from a relay network based on westend-local.
After some modification we finished an era with 7 targets, and a staking desired target of 10.

When trying to submit a solution for the election the `feasibility_check` fails because of this check:
```
ensure!(winners.len() as u32 == desired_targets, FeasibilityError::WrongWinnerCount);
```
https://github.com/paritytech/substrate/blob/master/frame/election-provider-multi-phase/src/lib.rs#L1367

I'm not sure if it is a valid check here or if this should check if the winners is not over the desired_targets only.
(cc @JoshOrndorff and @kianenigma )

### PR DESCRIPTION
**Pull request by georgesdib**
fixes #9478 

Amended some tests that were failing and won't fail after this change.

@kianenigma 

Polkadot address: 131dPecTmpTC1p1ofemufqFBJo9vNbV7dkgN7vWwKnaSMkC4

### CODE REVIEW

File: frame/election-provider-support/src/lib.rs
Comment (kianenigma) [inline frame/election-provider-support/src/lib.rs:230]: calling `targets` is super expensive, we can't afford to call it once here, and once in `pallet-election-provider-multi-phase`. 

What I meant by the call site of `ElectionDataProvider` is in fact `pallet-election-provider-multi-phase`. 

https://github.com/paritytech/substrate/blob/42b2d623d058197aebc3c737fb44fbbf278a85b4/frame/election-provider-multi-phase/src/lib.rs#L1359

File: frame/election-provider-multi-phase/src/lib.rs
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/lib.rs:1670]: this test should be renamed `desired_targets_gets_capped`, and we need a test that actually triggers the call path of `WrongWinnerCount`.

File: frame/election-provider-multi-phase/src/unsigned.rs
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/unsigned.rs:1047]: This test should still trigger the same error `PreDispatchWrongWinnerCount`.

You have two ways to do this: 

1. alter the solution manually to have too few winners.
2. alter the solution manually to have too many winners.

**Review by kianenigma — COMMENTED** [review]
Logic seems correct, tests need a bit more work. It was hard for me to give more advice by just looking at the diff since I don't quite recall the tests, but will try and clone the branch sometime.

File: frame/election-provider-multi-phase/src/lib.rs
Comment (georgesdib) [reply frame/election-provider-multi-phase/src/lib.rs:1667]: That is fixed, can you please let me know if this is what you had in mind to trigger `WrongWinnerCount`?
I have also removed what looks like an outdated comment around the line that triggers the `WrongWinnerCount` error.

File: frame/election-provider-multi-phase/src/unsigned.rs
Comment (georgesdib) [reply frame/election-provider-multi-phase/src/unsigned.rs]: Done, if you can please let me know if this is what you had in mind?

File: frame/election-provider-multi-phase/src/lib.rs
Comment (kianenigma) [inline frame/election-provider-multi-phase/src/lib.rs:1672]:
```suggestion
	fn less_than_desired_targets_fails() {
```

File: frame/election-provider-support/src/lib.rs
Comment (kianenigma) [inline frame/election-provider-support/src/lib.rs:228]: Sorry, I am a bit strict on documentation, and want to add a few suggestions here: 

1. Preferably don't reference `election-provider-multi-phase` here. The existence of this pallet should not be assumed here. 
2. Avoid first person plural, as it is not commonly used in the docs.
3. Describe what has to be done, or what is expected, not what our intentions or rationale are. A trait implementer need to know (foremost) what they must do.  

```suggestion
	/// A sensible implementation should use the minimum between this value and [`Self::targets().len()`], since desiring a winner set larger than candidates is not feasible.  
	///
	/// This is documented further in issue: <https://github.com/paritytech/substrate/issues/9478>
```

**Review by kianenigma — APPROVED** [review]
LGTM! Thanks!

**Review by emostov — APPROVED** [review]
LGTM - thanks! Just make sure to take a look at https://github.com/paritytech/substrate/pull/10905/files#r815887659

File: frame/election-provider-support/src/lib.rs
Comment (kianenigma) [inline frame/election-provider-support/src/lib.rs:222]: sorry, but this broke the line-width now :D

**Comment by emostov** [comment]
Just waiting until audited to merge.

**Comment by georgesdib** [comment]
> Just waiting until audited to merge.

good to merge?

**Comment by kianenigma** [comment]
audit ongoing and will be done soon. speculatively merging to keep the upstream PRs not in too much delay/pain.

**Comment by kianenigma** [comment]
bot merge

**Pull request merged**
